core/src/main/scala/pureconfig/ConfigCursor.scala

```diff
@@ -0,0 +1,273 @@
+   *
+   * @param idx the index of the config for which a cursor should be returned
+   * @return a `Right` with a cursor to the config at `idx` if such a config exists, a `Left` with a list of failures
+   *         otherwise.
+   */
+  def atIndex(idx: Int): Either[ConfigReaderFailures, ConfigCursor] = {
+    atIndexOrUndefined(idx) match {
+      case idxCur if idxCur.isUndefined => fail(KeyNotFound(idxCur.path, location, indexKey(idx), Set()))
+      case idxCur => Right(idxCur)
+    }
+  }
+
+  /**
+   * Returns a cursor to the config at a given index. An out of range index will return a cursor to an undefined value.
+   *
+   * @param idx the index of the config for which a cursor should be returned
+   * @return a cursor to the config at `idx` if such a config exists, a cursor to an undefined value otherwise.
+   */
+  def atIndexOrUndefined(idx: Int): ConfigCursor =
+    ConfigCursor(if (validIndex(idx)) value.get(idx) else null, indexKey(idx) :: pathElems)
+
+  /**
+   * Returns a cursor to the tail of the config list pointed to by this cursor if non-empty.
+   *
+   * @return a `Some` with the tail of the config list if the list is not empty, `None` otherwise.
+   */
+  def tailOption: Option[ConfigListCursor] = {
+    if (value.isEmpty) None
+    else {
+      val newValue = ConfigValueFactory.fromAnyRef(value.asScala.drop(1).asJava)
+        .withOrigin(value.origin)
+        .asInstanceOf[ConfigList]
+
+      Some(ConfigListCursor(newValue, pathElems, offset = offset + 1))
+    }
+  }
+
+  /**
+   * Returns a list of cursors to the elements of the config list pointed to by this cursor.
+   *
+   * @return a list of cursors to the elements of the config list pointed to by this cursor.
+   */
+  def list: List[ConfigCursor] =
+    value.asScala.toList.drop(offset).zipWithIndex.map { case (cv, idx) => ConfigCursor(cv, indexKey(idx) :: pathElems) }
+}
+
+/**
+ * A `ConfigCursor` pointing to a config object.
+ */
+case class ConfigObjectCursor(value: ConfigObject, pathElems: List[String]) extends ConfigCursor {
+
+  /**
+   * Returns whether the config object pointed to by this cursor is empty.
+   */
+  def isEmpty: Boolean = value.isEmpty
+
+  /**
+   * Returns the size of the config object pointed to by this cursor.
+   */
+  def size: Int = value.size
+
+  /**
+   * Returns the list of keys of the config object pointed to by this cursor.
+   */
+  def keys: Iterable[String] =
+    value.keySet.asScala
+
+  /**
+   * Returns a cursor to the config at a given key.
+   *
+   * @param key the key of the config for which a cursor should be returned
+   * @return a `Right` with a cursor to the config at `key` if such a config exists, a `Left` with a list of failures
+   *         otherwise.
+   */
+  def atKey(key: String): Either[ConfigReaderFailures, ConfigCursor] = {
+    atKeyOrUndefined(key) match {
+      case keyCur if keyCur.isUndefined => fail(KeyNotFound(keyCur.path, location, key, map.values.map(_.path)))
```

Comment:
I don't think building `KeyNotFound` like this would produce a valid list of candidate keys for the missing one, because we'll be comparing complete paths with a key at this level.

Also, I'm not sure if we should provide candidate keys in the error message for custom `ConfigReader`s that use the `atKey` method. What do you think?

Reply:
Sounds good. I'll pass only the keys themselves to `KeyNotFound`, since the full path of each candidate does not convey additional information when printing the error message.

I think that the error message is still useful even for custom `ConfigReader`s. While it points to a misconfigured product hint when it's not the case, I think that it still provides an important clue to the developer.